src/celeritas/optical/CerenkovPreGenerator.hh

```diff
@@ -0,0 +1,131 @@
+
+    auto optical_dist_data = pre_cerenkov(rng);
+    if (optical_dist_data)
+    {
+        CerenkovGenerator cerenkov_gen(... , optical_dist_data, ...);
+        cerenkov_gen(rng);
+    }
+ * \endcode
+ */
+class CerenkovPreGenerator
+{
+  public:
+    // Construct with particle and material data
+    inline CELER_FUNCTION
+    CerenkovPreGenerator(ParticleTrackView const& particle_view,
+                         NativeCRef<OpticalPropertyData> const& properties,
+                         NativeCRef<CerenkovData> const& shared,
+                         OpticalMaterialId mat_id,
+                         OpticalStepCollectorData const& optical_step);
+
+    // Return a populated distribution data for the Cerenkov Generator
+    template<class Generator>
+    inline CELER_FUNCTION OpticalDistributionData operator()(Generator& rng);
+
+  private:
+    ParticleTrackView const particle_view_;
+    OpticalMaterialId mat_id_;
+    OpticalStepCollectorData step_collector_data_;
+    CerenkovDndxCalculator dndx_calculator_;
+};
+
+//---------------------------------------------------------------------------//
+// INLINE DEFINITIONS
+//---------------------------------------------------------------------------//
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct from particle, material, and optical step collection data.
+ */
+CELER_FUNCTION CerenkovPreGenerator::CerenkovPreGenerator(
+    ParticleTrackView const& particle_view,
+    NativeCRef<OpticalPropertyData> const& properties,
+    NativeCRef<CerenkovData> const& shared,
+    OpticalMaterialId mat_id,
+    OpticalStepCollectorData const& optical_step)
+    : particle_view_(particle_view)
+    , mat_id_(mat_id)
+    , step_collector_data_(optical_step)
+    , dndx_calculator_(properties, shared, mat_id, particle_view.charge())
+{
+    CELER_EXPECT(mat_id_);
+    CELER_EXPECT(step_collector_data_);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Return an \c OpticalDistributionData object. If no photons are sampled, an
+ * empty object is returned and can be verified via its own operator bool.
+ *
+ * The number of photons is sampled from a Poisson distribution with a mean
+ * \f[
+   \langle n \rangle = \ell_\text{step} \frac{dN}{dx}
+ * \f]
+ * where \f$ \ell_\text{step} \f$ is the step length.
+ */
+template<class Generator>
+CELER_FUNCTION OpticalDistributionData
+CerenkovPreGenerator::operator()(Generator& rng)
+{
+    auto const& pre = step_collector_data_.points[StepPoint::pre];
+    auto const& post = step_collector_data_.points[StepPoint::post];
+    units::LightSpeed beta(real_type{0.5}
+                           * (pre.speed.value() + post.speed.value()));
+
+    // Sample number of photons from a Poisson distribution
+    real_type sampled_num_photons = PoissonDistribution<real_type>(
+        dndx_calculator_(beta) * step_collector_data_.step_length)(rng);
```

Comment:
Since dN/dx could be zero, I think we'll need to calculate it before sampling from the Poisson distribution and return early in that case.